google/cloud/internal/async_long_running.h

```diff
@@ -0,0 +1,173 @@
+
+/// Extracts the value (or error) from a completed long-running operation
+Status ExtractOperationResultImpl(
+    StatusOr<google::longrunning::Operation> op,
+    google::protobuf::Message& result,
+    absl::FunctionRef<bool(google::protobuf::Any const&)> validate_any,
+    std::string const& location);
+
+/**
+ * Extracts the value from a completed long-running operation.
+ *
+ * This helper is used in `AsyncLongRunningOperation()` to extract the value (or
+ * error) from a completed long-running operation.
+ */
+template <typename ReturnType>
+StatusOr<ReturnType> ExtractLongRunningResult(
+    StatusOr<google::longrunning::Operation> op, std::string const& location) {
+  ReturnType result;
+  auto status = ExtractOperationResultImpl(
+      std::move(op), result,
+      [](google::protobuf::Any const& any) { return any.Is<ReturnType>(); },
+      location);
+  if (!status.ok()) return status;
+  return result;
+}
+
+/**
+ * Asynchronously starts and polls a long-running operation.
+ *
+ * Long-running operations [aip/151] are used for API methods that take a
+ * significant amount of time to complete (think minutes, maybe an hour). The
+ * gRPC API returns a "promise" object, represented by the
+ * `google::longrunning::Operation` proto, and the application (or client
+ * library) should periodically poll this object until it is "done".
+ *
+ * In the C++ client libraries we represent these long-running operations by
+ * a member function that returns `future<StatusOr<ReturnType>>`. This function
+ * is a helper to implement these member functions. It first starts the
+ * operation using an asynchronous retry loop, and then starts an asynchronous
+ * loop to poll the operation until it completes.
+ *
+ * The promise can complete with an error, which is represented by a
+ * `google::cloud::Status` object, or with success and some `ReturnType` value.
+ * The application may also configure the "polling policy", which may stop the
+ * polling even though the operation has not completed.
+ *
+ * Library developers would use this function as follows:
+ *
+ * @code
+ * class BarStub {
+ *  public:
+ *   virtual future<StatusOr<google::longrunning::Operation>> AsyncFoo(
+ *     google::cloud::CompletionQueue& cq,
+ *     std::unique_ptr<grpc::ClientContext> context,
+ *     FooRequest const& request) = 0;
+ *
+ *   virtual future<StatusOr<google::longrunning::Operation>> AsyncGetOperation(
+ *     google::cloud::CompletionQueue& cq,
+ *     std::unique_ptr<grpc::ClientContext> context,
+ *     google::longrunning::GetOperationRequest const& request) = 0;
+ * };
+ * @endcode
+ *
+ * The corresponding `*ConnectionImpl` class would look as follows:
+ *
+ * @code
+ * class BarConnectionImpl : public BarConnection {
+ *  public:
+ *   // Using C++14 for exposition purposes. The implementation supports C++11.
+ *   future<StatusOr<FooResponse>> Foo(FooRequest const& request) override {
+ *     return google::cloud::internal::AsyncLongRunningLoop(
```

Comment:
Is `AsyncLongRunningLoop` the right name, or should this be `AsyncLongRunningOperation` ?

Reply:
Fixed, also fixed the file names to match.